### PR TITLE
Story 1

### PR DESCRIPTION
**Pull request by Harry-Johnston**

### CODE REVIEW

File: Style.css [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+a {
+    text-decoration: none;
+}
+
+.title_logo {
+    font-size: 3rem;
+    font-family: 'Mouse Memoirs', sans-serif;
+    color: #FFFFFF;
+
```
Comment (norayna) [inline Style.css]: blank line delete pls

File: Style.css [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+.nav_button{
+    color: #FFFFFF;
+}
+
+.nav_bar {
+    padding: 10px;
+    font-family: 'Roboto', sans-serif;
+    font-size:2rem;
```
Comment (norayna) [inline Style.css]: missing space: `font-size: 2rem;`

File: Style.css [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+.title_block {
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    text-align: center;
+    border-bottom: solid black 5px;
+}
+.title_logo{
```
Comment (norayna) [inline Style.css]: needs a space before {

File: Style.css [hunk 4]
```diff
@@ -0,0 +1,76 @@
+* {
```
Comment (norayna) [inline Style.css]: the file needs to be called `style.css`

File: index.php [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        <title>Training Log</title>
+    </head>
+    <body>
+        <header>
+            <nav class="nav_bar">
+                <a href="#" class="nav_button">Add a session</a>
+            </nav>
+            <div class="title_block">
+                <h1 class="title_logo">Chalk it up!</h1>
+                <h2 class="title_subheading">Check out your previous sessions below:</h2>
```
Comment (norayna) [inline index.php]: in terms of purpose on the page this is not really an `h2` because it's not the title of a subsection of the page, it's more of a unique paragraph for adding vibes, therefore i would change this to a `<p>`

File: trainingFunction.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                break;
+            }
+        }
+        $per_date_string .= "</div>";
+        $output_string .= $per_date_string;
+    }
+    return $output_string;
+}
+
+function getAllDates(array $all_workouts_array):array{
```
Comment (norayna) [inline trainingFunction.php]: a better name would be `getAllUniqueDates` to make it super clear that it removes repeats

File: trainingFunction.php [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        }
+        $exercise = $workout_item['exercise'];
+        $weight_added_kg = ($workout_item['weight_added_kg']===null) ? 0 : $workout_item['weight_added_kg'];
+        $comments = ($workout_item['comments']===null) ? '' : $workout_item['comments'];
+        $all_workouts_array[$key]['html'] = "<div class='exercise_container'><h3> $exercise <span class='weight_added_span'> | " . $weight_added_kg . "kg Added </span></h3><p>$comments</p></div>";
+    }
+    return $all_workouts_array;
+}
+
+function perDateOutput(array $all_workouts_array, array $date_array):string{
```
Comment (norayna) [inline trainingFunction.php]: i would rename this function to something like `displayAllWorkouts` because that would make it easier to understand what it does at a glance

File: trainingFunction.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                return '';
+            }
+            if ($workout_item['date'] === $date) {
+                $exercise_string = $workout_item['html'];
+                $per_date_string .= $exercise_string;
+                array_shift($all_workouts_array);
+            } else {
+                break;
+            }
+        }
```
Comment (norayna) [inline trainingFunction.php]: it would be good here to add a comment that explains why you've done the `array_shift` and `break` because if someone was reading the code they probably wouldn't understand all the context about removing unnecessary steps from the loops

File: trainingFunction.php [hunk 9]
```diff
@@ -0,0 +1,74 @@
+<?php
+
+function dbPull():array{
```
Comment (norayna) [inline trainingFunction.php]: rename this function to something like `pullAllWorkoutsFromDb`

File: trainingFunction.php [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    $connectionString = 'mysql:host=db; dbname=training_diary';
+    $dbUsername = 'root';
+    $dbpassword = 'password';
+    $db = new PDO ($connectionString, $dbUsername, $dbpassword);
```
Comment (norayna) [inline trainingFunction.php]: there should be no space after PDO: `new PDO($connectionString, $dbUsername, $dbpassword);`

File: trainingFunction.php [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+    $db->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
+
+    $queryString = 'SELECT * FROM 
+             `training_log`
+                 ORDER BY `date` DESC 
+                 LIMIT 10;'; //order by date descending
```
Comment (norayna) [inline trainingFunction.php]: why limit?

File: trainingFunction.php [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+function dbPull():array{
+    $connectionString = 'mysql:host=db; dbname=training_diary';
+    $dbUsername = 'root';
+    $dbpassword = 'password';
+    $db = new PDO ($connectionString, $dbUsername, $dbpassword);
+
+    $db->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
+
+    $queryString = 'SELECT * FROM 
+             `training_log`
```
Comment (norayna) [inline trainingFunction.php]: it would be a lot more conventional to name the table something like `workouts` - the db can be called whatever we like but the table generally will be like so:
users
tweets
bank_accounts
etc, you take the name of the singular thing (a workout) and you make it plural (workouts)

File: trainingFunction.php [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+            !array_key_exists('date', $workout_item) ||
+            !array_key_exists('exercise', $workout_item) ||
+            !array_key_exists('weight_added_kg', $workout_item) ||
+            !array_key_exists('comments', $workout_item)) {
+            return [];
+        }
+        $exercise = $workout_item['exercise'];
+        $weight_added_kg = ($workout_item['weight_added_kg']===null) ? 0 : $workout_item['weight_added_kg'];
+        $comments = ($workout_item['comments']===null) ? '' : $workout_item['comments'];
+        $all_workouts_array[$key]['html'] = "<div class='exercise_container'><h3> $exercise <span class='weight_added_span'> | " . $weight_added_kg . "kg Added </span></h3><p>$comments</p></div>";
```
Comment (norayna) [inline trainingFunction.php]: on this line, if there are no comments there will be an empty `<p>` tag:
```html
.....
</h3>
<p></p>
</div>
```

it is better if the `<p>` is not there when there's no comments

File: test/testTrainingFunction.php [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+                    'comments'=>'blah blah blah'
+                ],
+                [
+                    'date'=>'2022-04-06',
+                    'exercise'=>'Repeater',
+                    'difficulty'=>'easy',
+                    'weight_added_kg'=>10,
+                    'comments'=>'these are some words'
+                ]
+            ];
```
Comment (norayna) [inline test/testTrainingFunction.php]: it would be good to add another item here that has the same date as one of these ones, that way you are also testing that the functions removes duplicates

File: test/testTrainingFunction.php [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+        $date_array = [
+            '2022-07-10',
+            '2022-04-06'
+        ];
+        $result = perDateOutput($db_array, $date_array);
+        $expected = "<div class='date_container'><h2 class='date_text'>2022-07-10</h2><div class='exercise_container'><h3> Max Hang <span class='weight_added_span'> | 50kg Added </span></h3><p>blah blah blah</p></div></div><div class='date_container'><h2 class='date_text'>2022-04-06</h2><div class='exercise_container'><h3> Repeater <span class='weight_added_span'> | 10kg Added </span></h3><p>these are some words</p></div></div>";
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testPerDateOutputFailure_1(){
```
Comment (norayna) [inline test/testTrainingFunction.php]: no need for the `_1` on any of the test functions

File: test/testTrainingFunction.php [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+                    'weight_added_kg'=>10,
+                    'comments'=>'these are some words'
+                ]
+            ];
+        $result = getAllDates($db_array);
+        $expected = ['2022-07-10', '2022-04-06'];
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testGetAllDatesFailure_1(){
```
Comment (norayna) [inline test/testTrainingFunction.php]: it would be good here to add a comment that explains what the failure is and what the graceful handling is

File: test/testTrainingFunction.php [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+                'difficulty'=>'easy',
+                'weight_added_kg'=>10,
+                'comments'=>'these are some words',
+                'html'=>"<div class='exercise_container'><h3> Repeater <span class='weight_added_span'> | 10kg Added </span></h3><p>these are some words</p></div>"
+            ]
+        ];
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testAddHtmlToWorkoutsFailure_1(){
```
Comment (norayna) [inline test/testTrainingFunction.php]: again i would add a comment explaining what the failure is and what the graceful handling is

File: test/testTrainingFunction.php (hunk 15 above)
Comment (norayna) [reply test/testTrainingFunction.php]: also again i would add a comment explaining what the failure and graceful handling are

File: test/testTrainingFunction.php [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+            '2022-07-10',
+            '2022-04-06'
+        ];
+        $result = perDateOutput($db_array, $date_array);
+        $expected = "<div class='date_container'><h2 class='date_text'>2022-07-10</h2><div class='exercise_container'><h3> Max Hang <span class='weight_added_span'> | 50kg Added </span></h3><p>blah blah blah</p></div></div><div class='date_container'><h2 class='date_text'>2022-04-06</h2><div class='exercise_container'><h3> Repeater <span class='weight_added_span'> | 10kg Added </span></h3><p>these are some words</p></div></div>";
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testPerDateOutputFailure_1(){
+        $db_array = [
```
Comment (norayna) [inline test/testTrainingFunction.php]: rename this variable to `$all_workouts_array`, which is what it has been renamed to in the function anyway

File: trainingFunction.php [hunk 19]
```diff
@@ -0,0 +1,74 @@
+<?php
```
Comment (norayna) [inline trainingFunction.php]: rename this file (and the test file) to `functions.php`

File: index.php [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+        <link rel="preconnect" href="https://fonts.googleapis.com">
+        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
+        <link href="https://fonts.googleapis.com/css2?family=Mouse+Memoirs&family=Roboto:wght@300;700&display=swap" rel="stylesheet">
+        <title>Training Log</title>
+    </head>
+    <body>
+        <header>
+            <nav class="nav_bar">
+                <a href="#" class="nav_button">Add a session</a>
+            </nav>
```
Comment (norayna) [inline index.php]: it would be better to remove this and add it in story 2 because at the moment it doesn't do anything and we only want finished features on the `main` branch

File: index.php [hunk 21]
```diff
@@ -0,0 +1,30 @@
+<?php require_once 'trainingFunction.php';
+$all_workouts_array = dbPull();
+$date_array = getAllDates($all_workouts_array);
+$all_workouts_array = addHtmlToWorkouts($all_workouts_array);
+
+?>
+<html lang = "en">
```
Comment (norayna) [inline index.php]: there should be no spaces aroung the = : `<html lang="en">`

also missing `<!DOCTYPE html>`

File: Style.css [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+    font-family: 'Roboto', sans-serif;
+    font-size:2rem;
+    text-decoration: none;
+}
+
+.title_block {
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    text-align: center;
```
Comment (norayna) [inline Style.css]: this line makes no difference because the line above it centers the things

File: Style.css [hunk 23]
```diff
@@ -0,0 +1,76 @@
...
+}
+
+.weight_added_span{
+    font-weight: 100;
+}
+
+.date_container {
+    border: black solid 1px;
+    gap: 40px;
```
Comment (norayna) [inline Style.css]: this line is not doing anything

File: Style.css [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+    margin: 10px;
+    padding: 10px;
+    min-width: 300px;
```
Comment (norayna) [inline Style.css]: probably maybe change this to `width: 500px`;

File: Style.css [hunk 25]
```diff
@@ -0,0 +1,76 @@
+* {
+    box-sizing: border-box;
+}
+
+html {
+    font-family: 'Roboto', sans-serif;
+    scroll-behavior: smooth;
+    font-size: 10px;
```
Comment (norayna) [inline Style.css]: change to 16px (as discussed)

**Review by norayna — CHANGES_REQUESTED** [review]
changes requestedddd

File: trainingFunction.php (hunk 11 above)
Comment (Harry-Johnston) [reply trainingFunction.php]: removed

File: trainingFunction.php (hunk 13 above)
Comment (Harry-Johnston) [reply trainingFunction.php]: fixed

File: functions.php [hunk 26]
```diff
@@ -0,0 +1,73 @@
+<?php
+
+function pullAllWorkoutsFromDb():array{
```
Comment (nealioa) [inline functions.php]: Lay this out like this:
```
function pullAllWorkoutsFromDb(): array
{
```

File: functions.php [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+    $connectionString = 'mysql:host=db; dbname=training_diary';
+    $dbUsername = 'root';
+    $dbpassword = 'password';
+    $db = new PDO($connectionString, $dbUsername, $dbpassword);
+
+    $db->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_ASSOC);
+
+    $queryString = 'SELECT * FROM
```
Comment (nealioa) [inline functions.php]: Replace the * with the names of the columns you want (don't forget the backticks!).

It is not recommended to have * in production code.

File: functions.php [hunk 28]
```diff
@@ -0,0 +1,73 @@
...
+        $exercise = $workout_item['exercise'];
+        $weight_added_kg = ($workout_item['weight_added_kg']===null) ? 0 : $workout_item['weight_added_kg'];
+        $comments = ($workout_item['comments']===null) ? '' : '<p>' . $workout_item['comments'] . '</p>' ;
+        $all_workouts_array[$key]['html'] = "<div class='exercise_container'><h3> $exercise <span class='weight_added_span'> | " . $weight_added_kg . "kg Added </span></h3>$comments</div>";
+    }
+    return $all_workouts_array;
+}
+
+function displayAllWorkouts(array $all_workouts_array, array $date_array):string{
+     $output_string = '';
```
Comment (nealioa) [inline functions.php]: This is indented one space too far, i.e. not lined up with the foreach below.

File: functions.php [hunk 29]
```diff
@@ -0,0 +1,73 @@
...
+    foreach($date_array as $date) {
+        $per_date_string = "<div class='date_container'><h2 class='date_text'>$date</h2>";
+        foreach ($all_workouts_array as $workout_item) {
+            if (!is_array($workout_item) ||
```
Comment (nealioa) [inline functions.php]: Lay this out like this:
```
if (
    !is_array($workout_item) ||
```

File: functions.php [hunk 30]
```diff
@@ -0,0 +1,73 @@
...
+                !array_key_exists('date', $workout_item) ||
+                !array_key_exists('html', $workout_item)){
```
Comment (nealioa) [inline functions.php]: Lay this out like this:
```
    !array_key_exists('html', $workout_item)
) {
```

File: style.css [hunk 31]
```diff
@@ -0,0 +1,78 @@
...
+    font-family: 'Roboto', sans-serif;
+    font-size: 2rem;
+    text-decoration: none;
+}
+
+.title_block {
+    display: flex;
+    flex-direction: column;
+    align-items: center;
+    border-bottom: solid black 5px;
```
Comment (nealioa) [inline style.css]: Replace the colour name with a hex value.

File: style.css [hunk 32]
```diff
@@ -0,0 +1,78 @@
...
+    font-weight: 100;
+}
+
+.date_container {
+    border: black solid 1px;
+    margin: 10px;
+    padding: 10px;
+    width: 500px;
+    border-radius: 40px;
+    background-color: #e9e3f4;
```
Comment (nealioa) [inline style.css]: All the hex values should be uppercase or lowercase.

File: test/testTrainingFunction.php [hunk 33]
```diff
@@ -0,0 +1,183 @@
+<?php
+require '../functions.php';
+use PHPUnit\Framework\TestCase;
+class testTrainingFunction extends TestCase {
+
+    public function testGetAllUniqueDatesSuccess_getDates(){
```
Comment (nealioa) [inline test/testTrainingFunction.php]: Space between ) and {

**Review by nealioa — CHANGES_REQUESTED** [review]

**Review by nealioa — APPROVED** [review]
